cluster/images/controller-manager/Dockerfile

```diff
@@ -10,8 +10,42 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-FROM photon:2.0
+################################################################################
+##                               BUILD ARGS                                   ##
+################################################################################
+# This build arg allows the specification of a custom Golang image.
+ARG GOLANG_IMAGE=golang:1.12.6
 
-ADD vsphere-cloud-controller-manager /bin/
+# The distroless image on which the CPI manager image is built.
+#
+# Please do not use "latest". Explicit tags should be used to provide
+# deterministic builds. This image doesn't have semantic version tags, but
+# the fully-qualified image can be obtained by entering
+# "gcr.io/distroless/static:latest" in a browser and then copying the
+# fully-qualified image from the web page.
+ARG DISTROLESS_IMAGE=gcr.io/distroless/static@sha256:9b60270ec0991bc4f14bda475e8cae75594d8197d0ae58576ace84694aa75d7a
+
+# This build arg is the version to embed in the CPI binary
+ARG VERSION=unknown
+
+################################################################################
+##                              BUILD STAGE                                   ##
+################################################################################
+# Build the manager as a statically compiled binary so it has no dependencies
+# libc, muscl, etc.
+FROM ${GOLANG_IMAGE} as builder
+WORKDIR /build
+COPY go.mod go.sum ./
+COPY pkg/    pkg/
+COPY cmd/    cmd/
+ENV CGO_ENABLED=0
+RUN go build -a -ldflags='-w -s -extldflags=static -X main.version=${VERSION}' -o vsphere-cloud-controller-manager ./cmd/vsphere-cloud-controller-manager
 
-CMD ["/bin/vsphere-cloud-controller-manager"]
+################################################################################
+##                               MAIN STAGE                                   ##
+################################################################################
+# Copy the manager into the distroless image.
+FROM ${DISTROLESS_IMAGE}
+LABEL maintainer="Travis Rhoden <trhoden@vmware.com>"
+COPY --from=builder /build/vsphere-cloud-controller-manager /vsphere-cloud-controller-manager
+ENTRYPOINT [ "/vsphere-cloud-controller-manager" ]
```

Comment:
This change will have an impact on the manifests, they currently pass the whole path of the binary as `/bin/vsphere-cloud-controller-manager` followed by arguments.

I'm fine with moving this to `ENTRYPOINT` (it just makes sense) but we need to amend the documentation/manifests at the same time.

Reply:
Good catch on this one, that makes sense. In fact, I think we should talk about what we want... CSI actually uses `CMD ["/bin/vsphere-csi"]`, and then all of it's arguments are env vars, so there is no call to a `command` with arguments in the YAML at all.

CAPV uses `ENTRYPOINT`. I wasn't trying to standardize this portion of the Dockerfiles across the projects -- just use a multi-stage build that always used a known Golang image to build and use distroless if possible.

Reply:
yeah, I really wasn't intending to change this part... good catch, I'm going to change it back.

Reply:
I'm not against changing it to `ENTRYPOINT` as this would remove the need to specify a full path to the binary in the yaml file, and I see that as a desirable change 😄

Since the manifests need to be changed anyway (given the new location of the containers) maybe it's the right time to standardize on `ENTRYPOINT`